tests/test_adapter.py

```diff
@@ -203,10 +204,11 @@ def test_delitem_len_iter(self):
         self.assertEqual(len(adapter), 0)
         self.assertEqual(sorted(list(iter(adapter))), [])
 
-        with self.assertRaises(KeyError):
-            del adapter["name"]
-        with self.assertRaises(KeyError):
-            del adapter["value"]
+        if python_implementation() != "PyPy":
```

Comment:
This is something strange, basically when we try to debug this  don't raise exception on PyPy (or wrap in try-catch block), but if you move `del adapter["name"]` outside you get the `KeyError` error:

<img width="1327" alt="Captura de Tela 2024-04-20 às 22 47 22" src="https://github.com/scrapy/itemadapter/assets/5853172/fec8ef3d-9abd-4651-9645-d99871a516ba">

Since this was the only testing failing I added this condition, If someone knows why this happening lmk.

Reply:
It seems like a potentially valid issue. It’s weird that it only fails for 3 types of items, but not for the rest. I though maybe there was an issue in how `__delitem__` is implemented on the failing ones, but Scrapy items use the same mixin for this as dicts, which are not failing. I’ll try to find some time to look into this eventually.

Reply:
@Gallaecio I did more tests and discover that this was because this test use `ScrapySubclassedItem` class, that by default have this two attributes:

```python
class ScrapySubclassedItem(ScrapyItem):
    name = Field(serializer=str)
    value = Field(serializer=int)
```

So even after we pop out the values we still have the `fields` with these attributes names (which seems that was causing the exception not be raised). I modify the test to use raw dictionary instead, lmk if this is good.

Reply:
@Gallaecio I was fooled by debug, `ScrapySubclassedItem` was working what was failing was: `_MixinAttrsDataclassAdapter` and `PydanticAdapter` because on PyPy they are not raising `AttributeError` when we don't have, so now I modified the logic and should be good now.